File: ts/src/store.ts
import * as fs from 'fs';
import * as path from 'path';
import type { OpLog } from './oplog';
import type { RegionMap } from './region';
import type { FounderTree } from './founder';
import type { Operation, CodeRegion, OpStatus, Participant } from './types';

export class RadStore {
  private radDir: string;

  private constructor(radDir: string) {
    this.radDir = radDir;
  }

  static open(projectDir: string): RadStore {
    const radDir = path.join(projectDir, '.rad');
    if (!fs.existsSync(radDir)) {
      throw new Error('Not a rad project (run rad init first)');
    }
    return new RadStore(radDir);
  }

  loadOplog(): Operation[] {
    const oplogPath = path.join(this.radDir, 'oplog.json');
    if (!fs.existsSync(oplogPath)) {
      return [];
    }

    try {
      const content = fs.readFileSync(oplogPath, 'utf-8');
      return JSON.parse(content);
    } catch (e) {
      throw new Error(`error: corrupt or invalid oplog.json: ${e}`);
    }
  }

  saveOplog(operations: Operation[]): void {
    const oplogPath = path.join(this.radDir, 'oplog.json');
    const json = JSON.stringify(operations);
    fs.writeFileSync(oplogPath, json);
  }

  loadRegions(): CodeRegion[] {
    const regionsPath = path.join(this.radDir, 'regions.json');
    if (!fs.existsSync(regionsPath)) {
      return [];
    }

    try {
      const content = fs.readFileSync(regionsPath, 'utf-8');
      return JSON.parse(content);
    } catch {
      return [];
    }
  }

  saveRegions(regions: CodeRegion[]): void {
    const regionsPath = path.join(this.radDir, 'regions.json');
    const json = JSON.stringify(regions);
    fs.writeFileSync(regionsPath, json);
  }

  loadParticipants(): Participant[] {
    const participantsPath = path.join(this.radDir, 'participants.json');
    if (!fs.existsSync(participantsPath)) {
      return [];
    }

    try {
      const content = fs.readFileSync(participantsPath, 'utf-8');
      return JSON.parse(content);
    } catch {
      return [];
    }
  }

  saveParticipants(participants: Participant[]): void {
    const participantsPath = path.join(this.radDir, 'participants.json');
    const json = JSON.stringify(participants);
    fs.writeFileSync(participantsPath, json);
  }

  loadFounders(): { founders: Record<string, string>; rootFounder: string } {
    const configPath = path.join(this.radDir, 'config.json');
    let rootFounder = '';
    if (fs.existsSync(configPath)) {
      try {
        const content = fs.readFileSync(configPath, 'utf-8');
        const config = JSON.parse(content);
        rootFounder = config.founder || '';
      } catch {}
    }

    const foundersPath = path.join(this.radDir, 'founders.json');
    if (fs.existsSync(foundersPath)) {
      try {
        const content = fs.readFileSync(foundersPath, 'utf-8');
        const founders = JSON.parse(content);
        return { founders, rootFounder };
      } catch {}
    }

    return { founders: { '.': rootFounder }, rootFounder };
  }

  saveFounders(founders: Record<string, string>): void {
    const foundersPath = path.join(this.radDir, 'founders.json');
    const json = JSON.stringify(founders);
    fs.writeFileSync(foundersPath, json);
  }

  putSnapshot(filePath: string, content: string): void {
    const snapPath = path.join(this.radDir, 'snapshots', filePath);
    const dir = path.dirname(snapPath);
    fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(snapPath, content);
  }

  getSnapshot(filePath: string): string | null {
    const snapPath = path.join(this.radDir, 'snapshots', filePath);
    try {
      return fs.readFileSync(snapPath, 'utf-8');
    } catch {
      return null;
    }
  }

  compact(): void {
    const operations = this.loadOplog();

    // Group operations by file path and collect accepted writes
    const fileContents = new Map<string, string>();
    const acceptedIds = new Set<string>();

    for (const op of operations) {
      if (op.status === 'accepted' && op.op_type === 'Write') {
        // Extract file path from region_id (format: "file:start-end")
        const colonPos = op.region_id.indexOf(':');
        if (colonPos !== -1) {
          const filePath = op.region_id.substring(0, colonPos);
          fileContents.set(filePath, op.content);
          acceptedIds.add(op.id);
        }
      }
    }

    // Write snapshots
    for (const [filePath, content] of fileContents.entries()) {
      this.putSnapshot(filePath, content);
    }

    // Remove accepted operations from oplog
    const newOperations = operations.filter(op => !acceptedIds.has(op.id));
    this.saveOplog(newOperations);
  }
}

File: rust/src/store.rs
use std::path::{Path, PathBuf};
use std::fs;
use std::collections::HashMap;
use crate::oplog::OpLog;
use crate::region::RegionMap;
use crate::founder::FounderTree;
use crate::types::{Operation, OpStatus, CodeRegion};

pub struct RadStore {
    rad_dir: PathBuf,
}

impl RadStore {
    pub fn open(project_dir: &Path) -> Result<Self, String> {
        let rad_dir = project_dir.join(".rad");
        if !rad_dir.exists() {
            return Err("Not a rad project (run rad init first)".to_string());
        }
        Ok(Self { rad_dir })
    }

    pub fn load_oplog(&self) -> Result<OpLog, String> {
        let oplog_path = self.rad_dir.join("oplog.json");
        if !oplog_path.exists() {
            return Ok(OpLog::new());
        }

        let content = fs::read_to_string(&oplog_path)
            .map_err(|e| format!("error: failed to read oplog.json: {}", e))?;

        let ops = serde_json::from_str::<Vec<Operation>>(&content)
            .map_err(|e| format!("error: corrupt or invalid oplog.json: {}", e))?;

        let mut oplog = OpLog::new();
        for op in ops {
            oplog.add_operation(op);
        }
        Ok(oplog)
    }

    pub fn save_oplog(&self, oplog: &OpLog) -> Result<(), String> {
        let oplog_path = self.rad_dir.join("oplog.json");
        let ops = oplog.get_all_operations();
        let json = serde_json::to_string(&ops).map_err(|e| e.to_string())?;
        fs::write(&oplog_path, json).map_err(|e| e.to_string())?;
        Ok(())
    }

    pub fn load_regions(&self) -> RegionMap {
        let regions_path = self.rad_dir.join("regions.json");
        if !regions_path.exists() {
            return RegionMap::new();
        }

        match fs::read_to_string(&regions_path) {
            Ok(content) => {
                match serde_json::from_str::<Vec<CodeRegion>>(&content) {
                    Ok(regions) => {
                        let mut map = RegionMap::new();
                        for region in regions {
                            map.register(region);
                        }
                        map
                    }
                    Err(_) => RegionMap::new()
                }
            }
            Err(_) => RegionMap::new()
        }
    }

    pub fn save_regions(&self, region_map: &RegionMap) -> Result<(), String> {
        let regions_path = self.rad_dir.join("regions.json");
        let regions = region_map.get_all_regions();
        let json = serde_json::to_string(&regions).map_err(|e| e.to_string())?;
        fs::write(&regions_path, json).map_err(|e| e.to_string())?;
        Ok(())
    }

    pub fn load_founders(&self) -> FounderTree {
        let config_path = self.rad_dir.join("config.json");
        let root_founder = if config_path.exists() {
            if let Ok(content) = fs::read_to_string(&config_path) {
                if let Ok(config) = serde_json::from_str::<serde_json::Value>(&content) {
                    config.get("founder")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string()
                } else {
                    String::new()
                }
            } else {
                String::new()
            }
        } else {
            String::new()
        };

        let founders_path = self.rad_dir.join("founders.json");
        if founders_path.exists() {
            if let Ok(content) = fs::read_to_string(&founders_path) {
                return FounderTree::from_json(&content, &root_founder);
            }
        }
        FounderTree::new(&root_founder)
    }

    pub fn save_founders(&self, tree: &FounderTree) -> Result<(), String> {
        let founders_path = self.rad_dir.join("founders.json");
        let json = tree.to_json();
        fs::write(&founders_path, json).map_err(|e| e.to_string())?;
        Ok(())
    }

    pub fn put_snapshot(&self, file_path: &str, content: &str) -> Result<(), String> {
        let snap_path = self.rad_dir.join("snapshots").join(file_path);
        if let Some(parent) = snap_path.parent() {
            fs::create_dir_all(parent).map_err(|e| e.to_string())?;
        }
        fs::write(&snap_path, content).map_err(|e| e.to_string())?;
        Ok(())
    }

    pub fn get_snapshot(&self, file_path: &str) -> Option<String> {
        let snap_path = self.rad_dir.join("snapshots").join(file_path);
        fs::read_to_string(&snap_path).ok()
    }

    pub fn compact(&self) -> Result<(), String> {
        let mut oplog = self.load_oplog()?;
        let ops = oplog.get_all_operations();

        // Group operations by file path and collect accepted writes
        let mut file_contents: HashMap<String, String> = HashMap::new();
        let mut accepted_ids: Vec<String> = Vec::new();

        for op in ops {
            if op.status == OpStatus::Accepted && matches!(op.op_type, crate::types::OpType::Write) {
                // Extract file path from region_id (format: "file:start-end")
                if let Some(colon_pos) = op.region_id.find(':') {
                    let file_path = &op.region_id[..colon_pos];
                    file_contents.insert(file_path.to_string(), op.content.clone());
                    accepted_ids.push(op.id.clone());
                }
            }
        }

        // Write snapshots
        for (file_path, content) in file_contents {
            self.put_snapshot(&file_path, &content)?;
        }

        // Remove accepted operations from oplog
        for id in accepted_ids {
            oplog.remove_operation(&id);
        }

        self.save_oplog(&oplog)?;
        Ok(())
    }
}
